### ISSUE TITLE
Closing Context Manager

### ISSUE THREAD

**Issue by christian-intra2net** [decalage2/oletools#231]
Running unittests in python3 creates an awful lot of resource warnings. Some of them may be caused by all those `open(filename)` and `ZipFile.open(subfile)` calls that not always close() the resulting handles. That is easy to fix using `with [ZipFile.]open(arg) as file_handle: ...`.

But OleFileIO and OleStream also need to be close()ed after use and that does not always happen in the oletools code and almost never happens in error cases.

I was just about to wrap every OleFileIO use into a try-except block with a close() in the finally-block. But then I realized that Context Managers like the ones for file and ZipFile mentioned above would be much more pythonic. Instead of
```python
        ole = None
        try:
            ole = OleFileIO(args)
            do_something(ole)
        except Exception:
            raise
        finally:
            if ole is not None:
               ole.close()
```
you can do this:
```python
        with OleFileIO(args) as ole:
            do_something(ole)
```
That is easy enough, just needs a `__enter__` and `__exit__` method. We could
1. add this to olefile.OleFileIO itself, and/or
2. create a oletools/utils.py that implements a thin wrapper around OleFileIO, and/or
3. create a oletools/utils.py that implements a general AutoCloser context manager that calls close on whatever it gets, so the code would become `with AutoCloser(OleFileIO(args)) as ole: ...`, or
4. add all those try-except blocks to each call as described above (my least favourite), or
5. ignore the issue and the warnings (even more least favourite)

Similarly, there are many calls to `data=ole.openstream(name).read()` or `...getvalue()`. These also leave open stream handles. For 1. and 2. above we could add a `readstream()` method to replace that, for 3. above we could replace it with `with AutoCloser(ole.openstream(name)) as stream: data=stream.read()`. Also, we could repeat 1. or 2. 3. for `OleStream` to ensure closing of the other `openstream` calls.

Then,  there are these calls to `OleFileIO._open` that are requried to get orphaned streams. pylint hates these (_open is "protected"), and context managers for these would also be cool. We could:
1. add an optional argument `orphans=False` to `olefile.OleFileIO.listdir` to make pylint happy, and/or
2. overwrite listdir in a `oletools/utils.OleFileIOWrapper` class, and/or
3. create a `oletools/utils.listdir_with_orphans()`, and/or
4. use `oletools/utils.AutCloser` on them.

My favourite would be to modify olefile itself since these changes should not break any existing code. What do you think? I would be glad to implement any of these.

**Comment by decalage2** [comment]
Lots of good ideas indeed. The best would be to improve olefile itself. I plan to remove the thirdparty folder from oletools soon, and use setuptools dependencies instead to avoid maintaining copies of other repositories in oletools. If possible, I'd like this to be implemented in small chunks (if you submit PRs), to avoid breaking things. Thanks!

**Comment by decalage2** [comment]
Just a few notes: 
- we need to be careful not to close OleFileIO too soon, because an olefile object may be used in several places.
- OleStream is just a BytesIO/StringIO object, so not closing it immediately is not a big deal (I guess it's done by the garbage collector) unless we open lots of files and/or streams in a short time. But you're right it's much cleaner to close it explicitly.
- Are you sure a statement like `data=ole.openstream(name).read()` does not close the OleStream automatically, since the object is not stored in any variable?

**Comment by christian-intra2net** [comment]
No, I was not sure whether the file handle from `data=ole.openstream(name).read()` is closed automatically. I found [this stackoverflow question](https://stackoverflow.com/questions/38660609/when-how-does-an-anonymous-file-object-close) which basically says that it does get closed since the interpreter is usually clever enough, but that behaviour is not guaranteed. I guess that goes for all of the file-handle-closing: one should close them but if not nothing bad happens.

So I guess this whole thing is low-prio

**Comment by decalage2** [comment]
OK, fair enough. It's not urgent, but you're right that it would be much cleaner to use "with" and context managers.

**Comment by christian-intra2net** [comment]
Update:

- OleFileIO now has a closing context manager
- all ResourceWarnings about unclosed files (happen when the garbage collector closes an open file handle) have been addressed in PR #446
